pkg/polaris/aws/aws.go

```diff
@@ -426,49 +471,281 @@ func (a API) RemoveAccount(ctx context.Context, account AccountFunc, feature cor
 
-	return nil
+	return profiles, roles, nil
+}
+
+// AccountArtifacts returns the artifacts added to the cloud account.
+func (a API) AccountArtifacts(ctx context.Context, id IdentityFunc) (map[string]string, map[string]string, error) {
+	a.log.Print(log.Trace)
+
+	nativeID, err := a.toNativeID(ctx, id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	artifacts, err := aws.Wrap(a.client).ArtifactsToDelete(ctx, nativeID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s", err)
+	}
+
+	instanceProfiles := make(map[string]string)
+	roles := make(map[string]string)
+	var skipped []string
+	for _, artifact := range artifacts {
+		for _, artifact := range artifact.ArtifactsToDelete {
+			switch {
+			case strings.HasSuffix(artifact.ExternalArtifactKey, instanceProfileSuffix):
+				key := strings.TrimSuffix(artifact.ExternalArtifactKey, instanceProfileSuffix)
+				instanceProfiles[key] = artifact.ExternalArtifactValue
+			case strings.HasSuffix(artifact.ExternalArtifactKey, roleArnSuffix):
+				key := strings.TrimSuffix(artifact.ExternalArtifactKey, roleArnSuffix)
+				roles[key] = artifact.ExternalArtifactValue
+			default:
+				skipped = append(skipped, artifact.ExternalArtifactKey)
+			}
+		}
+	}
+	a.log.Printf(log.Debug, "Skipped the following artifacts: %v", skipped)
+
+	return instanceProfiles, roles, nil
+}
+
+// AddAccountArtifacts adds the specified artifacts, instance profiles and
+// roles, to the cloud account.
+func (a API) AddAccountArtifacts(ctx context.Context, id IdentityFunc, features []core.Feature, instanceProfiles map[string]string, roles map[string]string) (uuid.UUID, error) {
+	a.log.Print(log.Trace)
+
+	account, err := a.Account(ctx, id, core.FeatureAll)
+	if err != nil {
+		return uuid.Nil, err
+	}
+
+	externalArtifacts := make([]aws.ExternalArtifact, 0, len(instanceProfiles)+len(roles))
+	for key, value := range instanceProfiles {
+		if !strings.HasSuffix(key, instanceProfileSuffix) {
+			key = key + instanceProfileSuffix
+		}
+		externalArtifacts = append(externalArtifacts, aws.ExternalArtifact{
+			ExternalArtifactKey:   key,
+			ExternalArtifactValue: value,
+		})
+	}
+	for key, value := range roles {
+		if !strings.HasSuffix(key, roleArnSuffix) {
+			key = key + roleArnSuffix
+		}
+		externalArtifacts = append(externalArtifacts, aws.ExternalArtifact{
+			ExternalArtifactKey:   key,
+			ExternalArtifactValue: value,
+		})
+	}
+
+	var mappings []aws.NativeIDToRSCIDMapping
+	for i := 0; i < 7; i++ {
+		mappings, err = aws.Wrap(a.client).RegisterFeatureArtifacts(ctx, aws.Cloud(account.Cloud), []aws.AccountFeatureArtifact{{
+			NativeID:  account.NativeID,
+			Features:  features,
+			Artifacts: externalArtifacts,
+		}})
+		if err != nil {
+			return uuid.Nil, fmt.Errorf("failed to register feature artifacts: %s", err)
+		}
+		if len(mappings) != 1 {
+			return uuid.Nil, errors.New("expected account mappings for a single account")
+		}
+		if msg := mappings[0].Message; msg == "" || !strings.Contains(msg, "RBK30300003") {
```

Comment:
nit: Both the retry count (7) and the explicit error `RBK30300003` could be commented to add some more context to the reader.